lib/ansible/galaxy/collection/__init__.py

```diff
@@ -58,495 +129,93 @@
-
-            if not versions:
-                display.vvv("Collection '%s' is not available from server %s %s" % (collection, api.name,
-                                                                                    api.api_server))
-                continue
-
-            display.vvv("Collection '%s' obtained from server %s %s" % (collection, api.name, api.api_server))
-            break
-        else:
-            raise AnsibleError("Failed to find collection %s:%s" % (collection, requirement))
-
-        req = CollectionRequirement(namespace, name, None, api, versions, requirement, force, parent=parent,
-                                    metadata=galaxy_meta, allow_pre_releases=allow_pre_release)
-        return req
+    modified_content = []  # type: List[ModifiedContent]
+
+    # Verify the manifest hash matches before verifying the file manifest
+    expected_hash = _get_tar_file_hash(b_temp_tar_path, 'MANIFEST.json')
+    _verify_file_hash(b_collection_path, 'MANIFEST.json', expected_hash, modified_content)
+    manifest = _get_json_from_tar_file(b_temp_tar_path, 'MANIFEST.json')
+
+    # Use the manifest to verify the file manifest checksum
+    file_manifest_data = manifest['file_manifest_file']
+    file_manifest_filename = file_manifest_data['name']
+    expected_hash = file_manifest_data['chksum_%s' % file_manifest_data['chksum_type']]
+
+    # Verify the file manifest before using it to verify individual files
+    _verify_file_hash(b_collection_path, file_manifest_filename, expected_hash, modified_content)
+    file_manifest = _get_json_from_tar_file(b_temp_tar_path, file_manifest_filename)
+
+    # Use the file manifest to verify individual file checksums
+    for manifest_data in file_manifest['files']:
+        if manifest_data['ftype'] == 'file':
+            expected_hash = manifest_data['chksum_%s' % manifest_data['chksum_type']]
+            _verify_file_hash(b_collection_path, manifest_data['name'], expected_hash, modified_content)
+
+    if modified_content:
+        display.display(
+            'Collection {fqcn!s} contains modified content '
+            'in the following files:'.
+            format(fqcn=to_text(local_collection.fqcn)),
+        )
+        display.display(to_text(local_collection.fqcn))
+        display.vvv(to_text(local_collection.src))
+        for content_change in modified_content:
+            display.display('    %s' % content_change.filename)
+            display.vvv("    Expected: %s\n    Found: %s" % (content_change.expected, content_change.installed))
+        # FIXME: Why doesn't this raise a failed return code?
+    else:
+        display.vvv(
+            "Successfully verified that checksums for '{coll!s}' "
+            'match the remote collection'.
+            format(coll=local_collection),
+        )
 
 
 def build_collection(collection_path, output_path, force):
+    # type: (bytes, bytes, bool) -> Text
```

Comment:
For public APIs we should be accepting text strings and internally convert to bytes where needed. If we really must accept bytes (which in this case I don't see a reason to) we should prefix the arg with `b_` to indicate that.

The `b_` prefix may go away as we start to add type annotations but so far the behaviour is the standard we usually follow in ansible.

Reply:
Yeah, I guess it's not really public: I think it's not used outside this module.

Reply:
Addressed.